File: 2025-season6/7주차/최영권/ImplementQueueUsingStacks.java
package LinkedList;

import java.util.Stack;

public class ImplementQueueUsingStacks {
    Stack<Integer> inStack = new Stack<>();
    Stack<Integer> outStack = new Stack<>();

    public ImplementQueueUsingStacks() {

    }

    public static void main(String[] args) {
        ImplementQueueUsingStacks i = new ImplementQueueUsingStacks();
        i.push(1);
        i.push(2);
        System.out.println(i.pop());
        System.out.println(i.pop());

    }

    /**
     * 큐에 요소 추가
     * 시간복잡도: O(1)
     * - inStack에 push() 연산만 수행
     */
    public void push(int x) {
        inStack.push(x);
    }

    /**
     * 큐의 front 요소 제거 및 반환
     * 시간복잡도: O(1) amortized, 최악 O(n)
     * - outStack이 비어있으면: inStack의 모든 요소를 outStack으로 이동 → O(n)
     * - outStack에 요소가 있으면: 바로 pop() → O(1)
     * - 각 요소는 최대 2번(in→out, out→제거)만 이동되므로 amortized O(1)
     */
    public int pop() {
        if (outStack.isEmpty()) {
            while (!inStack.isEmpty()) {
                outStack.push(inStack.pop());
            }
            return outStack.pop();
        }
        return outStack.pop();
    }

    /**
     * 큐의 front 요소 조회 (제거하지 않음)
     * 시간복잡도: O(1) amortized, 최악 O(n)
     * - pop()과 동일한 로직, peek()만 다름
     * - outStack이 비어있으면 inStack 요소들을 이동 → O(n)
     * - outStack에 요소가 있으면 바로 peek() → O(1)
     */
    public int peek() {
        if (outStack.isEmpty()) {
            while (!inStack.isEmpty()) {
                outStack.push(inStack.pop());
            }
            return outStack.peek();
        }
        return outStack.peek();
    }

    /**
     * 큐가 비어있는지 확인
     * 시간복잡도: O(1)
     * - 두 스택의 empty() 체크만 수행
     */
    public boolean empty() {
        return outStack.empty() && inStack.empty();
    }
}

File: 2025-season6/7주차/이영범/implement-queue-using-stacks.ts
/*
232. Implement Queue using Stacks
Solved
Easy
Topics
premium lock icon
Companies
Implement a first in first out (FIFO) queue using only two stacks. The implemented queue should support all the functions of a normal queue (push, peek, pop, and empty).

Implement the MyQueue class:

void push(int x) Pushes element x to the back of the queue.
int pop() Removes the element from the front of the queue and returns it.
int peek() Returns the element at the front of the queue.
boolean empty() Returns true if the queue is empty, false otherwise.
Notes:

You must use only standard operations of a stack, which means only push to top, peek/pop from top, size, and is empty operations are valid.
Depending on your language, the stack may not be supported natively. You may simulate a stack using a list or deque (double-ended queue) as long as you use only a stack's standard operations.
 

Example 1:

Input
["MyQueue", "push", "push", "peek", "pop", "empty"]
[[], [1], [2], [], [], []]
Output
[null, null, null, 1, 1, false]

Explanation
MyQueue myQueue = new MyQueue();
myQueue.push(1); // queue is: [1]
myQueue.push(2); // queue is: [1, 2] (leftmost is front of the queue)
myQueue.peek(); // return 1
myQueue.pop(); // return 1, queue is [2]
myQueue.empty(); // return false
 

Constraints:

1 <= x <= 9
At most 100 calls will be made to push, pop, peek, and empty.
All the calls to pop and peek are valid.
 

Follow-up: Can you implement the queue such that each operation is amortized O(1) time complexity? In other words, performing n operations will take overall O(n) time even if one of those operations may take longer.
 */
export class MyStack<T> {
  data: T[];
  constructor() {
    this.data = [];
  }
  push(x: T) {
    this.data.push(x);
  }
  pop() {
    return this.data.pop();
  }
  peek() {
    return this.data[this.data.length - 1];
  }
  empty() {
    return this.data.length === 0;
  }
}

// 최종 TC: O(1) + Amortized O(1) + Amortized O(1) + O(1) = O(4) = O(1)
export class MyQueue {
  private inputStack: MyStack<number>;
  private outputStack: MyStack<number>;

  constructor() {
    this.inputStack = new MyStack<number>();
    this.outputStack = new MyStack<number>();
  }

  // Time: O(1)
  push(x: number): void {
    this.inputStack.push(x);
  }

  // TC: Amortized O(1)
  pop(): number {
    this.moveInputToOutput();
    return this.outputStack.pop()!;
  }

  // TC: Amortized O(1)
  peek(): number {
    this.moveInputToOutput();
    return this.outputStack.peek()!;
  }

  // TC: O(1)
  empty(): boolean {
    return this.inputStack.empty() && this.outputStack.empty();
  }

  /**
   * 헬퍼 메서드: inputStack의 모든 요소를 outputStack으로 이동
   * 순서가 뒤집혀서 가장 오래된 요소가 먼저 접근 가능하게 됨
   * Time: O(n), but amortized O(1) per operation
   */
  private moveInputToOutput(): void {
    // outputStack이 비어있을 때만 이동
    if (this.outputStack.empty()) {
      while (!this.inputStack.empty()) {
        this.outputStack.push(this.inputStack.pop()!);
      }
    }
  }
}
